src/core/metadata/qgslayermetadata.cpp

```diff
@@ -273,6 +278,312 @@ void QgsLayerMetadata::readFromLayer( const QgsMapLayer *layer )
+    oneLink.url = mne.attribute( QStringLiteral( "url" ) );
+    oneLink.description = mne.attribute( QStringLiteral( "description" ) );
+    oneLink.format = mne.attribute( QStringLiteral( "format" ) );
+    oneLink.mimeType = mne.attribute( QStringLiteral( "mimeType" ) );
+    oneLink.size = mne.attribute( QStringLiteral( "size" ) );
+    addLink( oneLink );
+  }
+
+  return true;
+}
+
+bool QgsLayerMetadata::writeMetadataXml( QDomElement &metadataElement, QDomDocument &document ) const
+{
+  // identifier
+  QDomElement identifier = document.createElement( QStringLiteral( "identifier" ) );
+  QDomText identifierText = document.createTextNode( mIdentifier );
+  identifier.appendChild( identifierText );
+  metadataElement.appendChild( identifier );
+
+  // parent identifier
+  QDomElement parentIdentifier = document.createElement( QStringLiteral( "parentidentifier" ) );
+  QDomText parentIdentifierText = document.createTextNode( mParentIdentifier );
+  parentIdentifier.appendChild( parentIdentifierText );
+  metadataElement.appendChild( parentIdentifier );
+
+  // language
+  QDomElement language = document.createElement( QStringLiteral( "language" ) );
+  QDomText languageText = document.createTextNode( mLanguage );
+  language.appendChild( languageText );
+  metadataElement.appendChild( language );
+
+  // type
+  QDomElement type = document.createElement( QStringLiteral( "type" ) );
+  QDomText typeText = document.createTextNode( mType );
+  type.appendChild( typeText );
+  metadataElement.appendChild( type );
+
+  // title
+  QDomElement title = document.createElement( QStringLiteral( "title" ) );
+  QDomText titleText = document.createTextNode( mTitle );
+  title.appendChild( titleText );
+  metadataElement.appendChild( title );
+
+  // abstract
+  QDomElement abstract = document.createElement( QStringLiteral( "abstract" ) );
+  QDomText abstractText = document.createTextNode( mAbstract );
+  abstract.appendChild( abstractText );
+  metadataElement.appendChild( abstract );
+
+  // keywords
+  QMapIterator<QString, QStringList> i( keywords() );
+  while ( i.hasNext() )
+  {
+    i.next();
+    QDomElement keywordsElement = document.createElement( QStringLiteral( "keywords" ) );
+    keywordsElement.setAttribute( QStringLiteral( "vocabulary" ), i.key() );
+    for ( const QString kw : i.value() )
+    {
+      QDomElement keyword = document.createElement( QStringLiteral( "keyword" ) );
+      QDomText keywordText = document.createTextNode( kw );
+      keyword.appendChild( keywordText );
+      keywordsElement.appendChild( keyword );
+    }
+    metadataElement.appendChild( keywordsElement );
+  }
+
+  // fees
+  QDomElement fees = document.createElement( QStringLiteral( "fees" ) );
+  QDomText feesText = document.createTextNode( mFees );
+  fees.appendChild( feesText );
+  metadataElement.appendChild( fees );
+
+  // constraints
+  for ( const QgsLayerMetadata::Constraint constraint : constraints() )
```

Comment:
This will detach - in this case you can use

    for ( const QgsLayerMetadata::Constraint& constraint : mConstraints )

since mConstraints is already const (as the method is const). Also note missing reference on `constraint`.